### PR TITLE
Deprecate old style BIO callback calls

### PR DESCRIPTION
**Pull request by t8m**
New style BIO_debug_callback_ex() function added to provide
replacement for BIO_debug_callback().

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
It would be really nice to deprecate this as the old style callbacks are fairly broken in regards to length type handling and we should remove them as soon as possible.

**Review by slontis — COMMENTED** [review]
Any deprecations should be in the migration_guide .
Search for
=head3 Deprecated function mappings

**Comment by t8m** [comment]
Added a note in migration guide. Rebased to fix trivial conflict in libcrypto.num.

**Review by paulidale — APPROVED** [review]

File: test/bio_callback_test.c [hunk 1]
```diff
@@ -20,7 +21,168 @@ static const char *my_param_argp[MAXCOUNT];
...
+
+    BIO_set_callback_ex(bio, my_bio_cb_ex);
+    i = BIO_write(bio, test1, test1len);
+    if (!TEST_int_eq(i, test1len)
+            || !TEST_int_eq(my_param_count, 2)
+            || !TEST_ptr_eq(my_param_b[0], bio)
+            || !TEST_int_eq(my_param_oper[0], BIO_CB_WRITE)
+            || !TEST_ptr_eq(my_param_argp[0], test1)
+            || !TEST_size_t_eq(my_param_len[0], test1len)
+            || !TEST_long_eq(my_param_argl[0], 0L)
```
Comment (slontis) [inline test/bio_callback_test.c:70]: Not sure it matter but you dont test argi here..

Comment (t8m) [reply test/bio_callback_test.c:70]: The argi is relevant only for BIO_CB_CTRL and the processed is relevant only when BIO_CB_RETURN is set. I can add a test case for the BIO_CB_CTRL.

**Comment by t8m** [comment]
> So argi was used to pass the length in the old callback? Does it even need argi in the new callback

Yes, for BIO_CB_CTRL.

**Comment by t8m** [comment]
Added a testcase for BIO_CB_CTRL. @paulidale please reconfirm. I do not think the merge waiting period needs to be restarted.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
As the fixup just added additional testing. I'm merging.

**Comment by t8m** [comment]
Merged. Thank you for the reviews.

**Pull request closed without merging**